runtime/value.go

```diff
@@ -0,0 +1,355 @@
+	if v.status == valueClosed {
+		v.Unlock()
+		return errValueClosed
+	}
+	if v.watch != nil {
+		v.watch.Close()
+	}
+	v.watch = nil
+	v.status = valueClosed
+	v.Unlock()
+
+	close(v.closedCh)
+	v.updateWg.Wait()
+
+	return nil
+}
+
+func (v *value) watchUpdates(watch kv.ValueWatch) {
+	defer v.updateWg.Done()
+
+	for {
+		select {
+		case _, ok := <-watch.C():
+			if !ok {
+				return
+			}
+			if err := v.updateFn(watch.Get()); err != nil {
+				v.log.Errorf("error updating rules: %v", err)
+			}
+		case <-v.closedCh:
+			return
+		}
+	}
+}
+
+func (v *value) initUpdate(update kv.Value) error {
+	latest, err := v.unmarshalFn(update)
+	if err != nil {
+		return err
+	}
+	newVersion := v.versionFn(latest)
+
+	// If this is the first time we update the rules, we need
+	// to go back in versions until we have found a version
+	// that's already in effect (or have fetched all versions).
+	var (
+		now         = v.nowFn()
+		currVersion = newVersion - 1
+		currCutover = v.cutoverFn(latest)
+		multiErr    = xerrors.NewMultiError()
+		historical  []interface{}
+	)
+
+	for ; currVersion > defaultVersion && currCutover.After(now); currVersion-- {
+		values, err := v.store.History(v.key, currVersion, currVersion+1)
+		if err != nil {
+			err = fmt.Errorf("error getting history for version %d: %v", currVersion, err)
+			multiErr = multiErr.Add(err)
+			continue
+		}
+
+		if len(values) != 1 {
+			err = fmt.Errorf("expected 1 value for version %d but got %d", currVersion, len(values))
+			multiErr = multiErr.Add(err)
+			continue
+		}
+
+		value, err := v.unmarshalFn(values[0])
+		if err != nil {
+			err = fmt.Errorf("error unmarshaling value for version %d: %v", currVersion, err)
+			multiErr = multiErr.Add(err)
+			continue
+		}
+
+		historical = append(historical, value)
+		now = v.nowFn()
+		currCutover = v.cutoverFn(value)
+	}
+
+	// Process updates in reverse order, oldest update first.
+	for i := len(historical) - 1; i >= 0; i-- {
+		v.processFn(historical[i])
+	}
+	v.processFn(latest)
+
+	v.Lock()
+	v.initialized = true
+	v.version = newVersion
+	v.Unlock()
+
+	return multiErr.FinalError()
+}
+
+func (v *value) cachedUpdate(update kv.Value) error {
+	latest, err := v.unmarshalFn(update)
+	if err != nil {
+		return err
+	}
+	newVersion := v.versionFn(latest)
+
+	v.RLock()
+	cachedVersion := v.version
+	v.RUnlock()
+
+	// If the cached version is higher than the new version,
+	// we skip the update.
+	if newVersion <= cachedVersion {
+		return nil
+	}
+
+	// If there are multiple versions in between the cached version
+	// and the new version, we need to retrieve every version between
+	// the latest cached version and the new version.
+	multiErr := xerrors.NewMultiError()
+	if newVersion > cachedVersion+1 {
+		values, err := v.store.History(v.key, cachedVersion+1, newVersion)
+		if err != nil {
+			err = fmt.Errorf("error getting history from version %d to version %d: %v", cachedVersion+1, newVersion, err)
+			multiErr = multiErr.Add(err)
```

Comment:
Again: do we need retry here?  I worry that if there is ever an error we'll just always have skipped it and never recover from it for the lifetime of the entire collector (until it's restarted).